### ISSUE TITLE
unable to handle brightness / media / hotkeys

### ISSUE THREAD

**Issue by briceburg** [rvaiya/keyd#638]
:wave: hello and happy new year! 

I would like to respond to `brightnessup` and `brightnessdown` keypresses, e.g.

```
# /etc/keyd/default.conf

brightnessdown = command(brightness.sh down)
brightnessup = command(brightness.sh up)
```

so that I can alter the brightness from a virtual console _or_ X11/Wayland...

however pressing the keys (Fn+f4 and Fn+f5 on my asus zenbook 14x laptop) does not appear to register a keyd event.

`keyd monitor` does not output anything when the keys are pressed, but it does for the other media keys like volumeup.

what's interesting is that the keypresses do appear to `showkey`, e.g.

```
# showkey
kb mode was unicode
[ if you are trying this under X, it might not work
since the X server is also reading /dev/console ]

press any key (program terminates 10s after last keypress)...
keycode  28 release
keycode 224 press
keycode 224 release
keycode 225 press
keycode 225 release
```

Is there a mechanism preventing the 224/225 keypresses from being handled by keyd? Thanks!

**Comment by briceburg** [comment]
My hunch is this may be HID related and `keyd` is unable to respond to certain events that appear to xev/xbindkeys?

There appear to be 3 keyboards registered at boot;

```
dmesg | grep input | grep -i keyboard
[    2.057455] input: AT Translated Set 2 keyboard as /devices/platform/i8042/serio0/input/input3
[    5.265941] input: ASUE140D:00 04F3:31B9 Keyboard as /devices/pci0000:00/0000:00:15.0/i2c_designware.0/i2c-15/i2c-ASUE140D:00/0018:04F3:31B9.0001/input/input10
[    5.787053] input: ASUE140D:00 04F3:31B9 Keyboard as /devices/pci0000:00/0000:00:15.0/i2c_designware.0/i2c-15/i2c-ASUE140D:00/0018:04F3:31B9.0001/input/input23
```

04F3:31B9 appears on input10 and input23. Is this normal?

Below is output from `keyd --monitor` showing the devices it listens to;

```sh
keyd --monitor
device added: 0001:0001 AT Translated Set 2 keyboard (/dev/input/event3)
device added: 04f3:31b9 ASUE140D:00 04F3:31B9 Mouse (/dev/input/event9)
device added: 04f3:31b9 ASUE140D:00 04F3:31B9 Touchpad (/dev/input/event10)
device added: 04f3:31b9 ASUE140D:00 04F3:31B9 Keyboard (/dev/input/event18)
device added: 04f3:40c3 ELAN9008:00 04F3:40C3 (/dev/input/event13)
device added: 04f3:40c3 ELAN9008:00 04F3:40C3 Stylus (/dev/input/event16)
AT Translated Set 2 keyboard    0001:0001       enter up
```

Thanks!

**Comment by briceburg** [comment]
As a followup, I found;
* The brightness keys are broadcast by the "Video Bus" device (id=7 above)
* The mic mute button is broadcast by the "Asus WMI hotkeys" (id=16 above).

To verify this, I used `xinput test` while pressing the brightness keys, e.g.

```sh
# xinput test 7
key press   232
key release 232
key press   233
key release 233
```

Please excuse my hackish discovery ways here, this is new frontier for me =)

So I went looking into keyd source and learned I could get verbose debug output via;

```
KEYD_DEBUG=2 bin/keyd monitor
...
DEBUG: src/device.c:129: capabilities of /dev/input/event4 (Video Bus): 0
```

meaning that the Video Bus device is being ignored (`resolve_device_capabilities` is returning a 0). 

Forcibly adding it via something like

```
if (strcmp("Video Bus",dev->name) == 0) {
  capabilities = 4;
}
```

seems to fix the issue :fireworks: ;

```
# KEYD_DEBUG=2 bin/keyd monitor
DEBUG: src/keyd.c:230: Debug mode activated
...
DEBUG: src/device.c:125: capabilities of /dev/input/event4 (Video Bus): 1
DEBUG: src/device.c:125: capabilities of /dev/input/event5 (Intel HID events): 0
DEBUG: src/device.c:125: capabilities of /dev/input/event3 (AT Translated Set 2 keyboard): 4
...
device added: 0001:0001 AT Translated Set 2 keyboard (/dev/input/event3)
device added: 0000:0006 Video Bus (/dev/input/event4)
device added: 04f3:31b9 ASUE140D:00 04F3:31B9 Mouse (/dev/input/event9)
device added: 04f3:31b9 ASUE140D:00 04F3:31B9 Touchpad (/dev/input/event11)
device added: 04f3:31b9 ASUE140D:00 04F3:31B9 Keyboard (/dev/input/event13)
device added: 04f3:40c3 ELAN9008:00 04F3:40C3 (/dev/input/event14)
device added: 04f3:40c3 ELAN9008:00 04F3:40C3 Stylus (/dev/input/event17)
DEBUG: src/device.c:470: unrecognized evdev event type: 4 4 28
DEBUG: src/device.c:459: key enter up
AT Translated Set 2 keyboard    0001:0001       enter up
DEBUG: src/device.c:459: key brightnessdown down
Video Bus       0000:0006       brightnessdown down
DEBUG: src/device.c:459: key brightnessdown up
Video Bus       0000:0006       brightnessdown up
DEBUG: src/device.c:459: key brightnessup down
Video Bus       0000:0006       brightnessup down
DEBUG: src/device.c:459: key brightnessup up
Video Bus       0000:0006       brightnessup up
```

although I must admit I am out of water. 

@rvaiya does it make sense to try to register inputs from the Video Bus device (brightness) and Asus WMI Hotkeys (Mic)? Or is it not really part of the scope of this project considering keyd runs as root and `command` mapping is bad. I always can fallback to xbindkeys but was really hoping on something lower level that works in the virtual console as well as X/Wayland.

**Comment by rvaiya** [comment]
Nice work :). Sorry for the belated response. The PR looks like the correct fix and should be merged soon.

**Comment by briceburg** [comment]
closed via fd0b99d73297cf255994a0bc414100bc7a6240f1
